### ISSUE TITLE
v5: Modal focus changes

### ISSUE THREAD

**Issue by mdo** [twbs/bootstrap#28481]
Skipping some previous v4-based PRs given our move to regular JS has already landed in `master`.

- #25277: Modal tabbing and shift+tabbing behavior
- #27646: Allow shift-tab focusing for modals
- #25651: Modal input autofocus attr & keyboard nav
- #21661: shift+tab without focused element within modal
- #27241: SHIFT + TAB should cycle backwards

/cc @Johann-S

**Comment by tmorehouse** [comment]
A simple way to "trap" the regular <kbd>TAB</kbd> and <kbd>SHIFT</kbd>+<kbd>TAB</kbd> is to  trick the browser into doing all the work figuring out which element to tab into when <kbd>SHIFT</kbd>+<kbd>TAB</kbd>.

The trick is to create two `<div tabindex="0" />` elements, one on top of the modal (before the header), and one after the footer.  Both div's are `0px` height.

When someone is in the first tabable element in the modal (i.e. it could be the `x` button), and they press <kbd>SHIFT</kbd>+<kbd>TAB</kbd>, the first tab trap `<div>` (the one before the header) receives a `focusin` event.  You trap that event, and then move focus to the on the bottom tab trap `<div>` (below the footer) and then trigger <kbd>SHIFT</kbd>+<kbd>TAB</kbd> keyboard event on the `<div>`.  The browser then kicks in and moves focus to the bottom most tabbable in the modal (assuming all the tabbables in the modal are default tabindex (i.e. `0`).

To get around having tabbables with higher than `0` tab index, one could just give the bottom tab trap `<div>` a rather high tabindex (i.e. `9999999999999`), or query the modal for `[tabindex]` elements and find the highest value and then set the bottom tab trap `<div>` to that computed tab index.

So there would be no need to do any crazy searching/sorting/computing to determine which element to tab to when the <kbd>SHIFT</kbd>+<kbd>TAB</kbd> happens on the first tabbable inside the modal.

**Comment by RyanBerliner** [comment]
Lots of mentions about `TAB` and `SHIFT + TAB` behavior in a couple issues, this seems to be the most current so hopefully its the correct spot.

I also agree that `SHIFT + TAB` should jump focus to the last focusable element in the modal when pressed with focus on the first focusable element. The recommended design pattern and keyboard support put out by w3c explains this as well. https://www.w3.org/TR/wai-aria-practices-1.1/examples/dialog-modal/dialog.html

Would this be a welcomed change? Or is there a reason behind its current implementation?

The behavior I've described is present in v4, and v5, as seen on the doc pages.
https://getbootstrap.com/docs/4.5/components/modal/
https://v5.getbootstrap.com/docs/5.0/components/modal/

**Comment by SandHilt** [comment]
Why don't use dialog tag element?

https://html.spec.whatwg.org/multipage/interactive-elements.html#the-dialog-element

**Comment by patrickhlauke** [comment]
because it's poorly supported and we'd need to polyfill it? https://caniuse.com/?search=dialog

**Comment by RyanBerliner** [comment]
Yeah, support for `<dialog>` is not great at the moment. Even in chrome, which claims support, native keyboard navigation on `<dialog open>` elements doesn't trap focus, or even screen reader navigation (I've tested voice over on chrome) 🤷‍♂️ .

**Comment by mdo** [comment]
What's the best path forward here @twbs/js-review @patrickhlauke? I added a third modal focus issue to the above, but I'm unclear what if anything we should do here.

**Comment by patrickhlauke** [comment]
i'd need to start digging into our code, but essentially we need to tweak current focus trapping/cycling behaviour.

in an ideal world, we'd be using `<dialog>` and polyfilling/shimming it, and/or using `inert` attribute and polyfilling/shimming it. but i believe we decided not to go down the polyfill route.

i've recently messed around with https://github.com/edenspiekermann/a11y-dialog (h/t @HugoGiraudel), so perhaps some bits from there could be of use. but until I had time to do a proper review of how our modal currently operates, it's tough to say. hoping to be able to carve out some time this coming week for this...

(and yes, agree this should only be done in `v5`, rather than trying to cram it into `v4`)

**Comment by HugoGiraudel** [comment]
Thank you for the shoutout @patrickhlauke!

[a11y-dialog](https://github.com/edenspiekermann/a11y-dialog) is small, spec-compliant, dependency-free and provides both a JS and a DOM API. It can also leverage the `<dialog>` HTML element although that is currently not recommended because the support across browser is inconsistent at best. Oh, and it supports dialogs (`role="dialog"`) *and* modals (`role="alertdialog"`) which is nice.

Let me know if I can help somewhat. :)

**Comment by patrickhlauke** [comment]
so, the current behaviour is due to the fairly naive `_enforceFocus()` function https://github.com/twbs/bootstrap/blob/main/js/src/modal.js#L281

```
_enforceFocus() {
  EventHandler.off(document, EVENT_FOCUSIN) // guard against infinite focus loop
  EventHandler.on(document, EVENT_FOCUSIN, event => {
    if (document !== event.target &&
        this._element !== event.target &&
        !this._element.contains(event.target)) {
      this._element.focus()
    }
  })
}
```

in essence, as soon as focus moves out of the modal, it sets focus back on the modal. what happens then if a user progresses forward through the modal using `Tab` is that once they get past the last focusable element inside the modal, their focus falls out, and focus is set back to the dialog container. so, if they now try to navigate backwards `Shift`+`Tab` they once again fall out of the modal, and focus is forced back to the modal.

conceptually, unless we want to go for a more abstracted/systemic solution such as `inert` (which currently needs to be polyfilled in all browsers https://github.com/wicg/inert), or using an actual `<dialog>` (which also needs to be polyfilled), one hacky solution is to use two empty listener elements - one at the start of the modal, one at the end. make them actually keyboard-focusable (`tabindex="0"`). when the one at the start of the modal gets focus, programmatically move focus to the last focusable element in the modal. vice-versa, if the listener element at the end of the modal gets focus, move focus programmatically to the first focusable element in the modal. not exactly elegant.

another option would be to add extra explicit keyboard listener to the modal container and detect if the specific `Shift`+`Tab` key combination is pressed when the outer modal container has focus, and if so programmatically move focus to the last focusable element in the modal. Of course this is not perfect, as it won't catch other ways in which focus may be moved, but should address the most common scenario.

**Comment by patrickhlauke** [comment]
> To get around having tabbables with higher than `0` tab index, one could just give the bottom tab trap `<div>` a rather high tabindex (i.e. `9999999999999`)

problem with that is that if there *are* no other elements with `tabindex` > 0, browsers will still prioritise the `<div>` with that super-high `tabindex` before cycling to the ones without any `tabindex`.

**Comment by HugoGiraudel** [comment]
In case this helps, on a11y-dialog, I handle focus trapping like this:

https://github.com/edenspiekermann/a11y-dialog/blob/824cf9c1a818a629007081b9b786ff846aec4b1b/a11y-dialog.js#L207

https://github.com/edenspiekermann/a11y-dialog/blob/824cf9c1a818a629007081b9b786ff846aec4b1b/a11y-dialog.js#L305-L319

https://github.com/edenspiekermann/a11y-dialog/blob/824cf9c1a818a629007081b9b786ff846aec4b1b/a11y-dialog.js#L416-L436

**Comment by RyanBerliner** [comment]
This is _not_ a pull request candidate in any way, just putting some of the ideas together to see what would actually need to be changed once you start getting your hands dirty.

1. You can't figure out whether someone is navigating backwards or forwards without either listening to the keyboard, or tracking navigation history and comparing to dom order... which seems like a pain. So listening to the keyboard seems like the way to go, and needs to be added. The only keyboard listener currently there is for dismissing the dialog.
2. To avoid the `tabindex` issues of putting dummy elements at the beginning and end of the modal, there needs to be a way to find all focusable elements inside the modal, so you know where the beginning and end is.

This is basically exactly what @HugoGiraudel outlined above.

Fitting that into the current bootstrap modal setup would either mean replacing the [`_enforceFocus()`](https://github.com/twbs/bootstrap/blob/main/js/src/modal.js#L281) event handling completely with a keyboard listener, or adding a keyboard listener alongside it. 

[Here is a proof of concept adding a keyboard listener](https://github.com/RyanBerliner/bootstrap/commit/1747b86a55a02f167f1cc3d28ea6280aa0783532) along side the (slightly modified) `focusin.bs.modal` event. In my mind that's an option, although a cleaner option might be just changing the existing `enforceFocus()` to get the events it needs.

**Comment by HugoGiraudel** [comment]
Regarding getting focusable elements, you might want to consider this: https://github.com/HugoGiraudel/focusable-selectors.

**Comment by RyanBerliner** [comment]
I'd be glad to put in some time to submit a pr for this - first though I'd like to make sure that the bootstrap team is on board with doing this for v5, specifically following [this proof of concept](https://github.com/RyanBerliner/bootstrap/commit/1747b86a55a02f167f1cc3d28ea6280aa0783532) as a roadmap for implementing the desired behavior. I'd imagine as a follow up pr to resolving this issue, the same behavior should be inherited into the new Offcanvas component, using a similar inheritance pattern as the relationship between tooltips and popovers.
